common/check_internal.h

```diff
@@ -54,6 +68,13 @@ template <TemplateString Kind, TemplateString File, int Line,
 
 }  // namespace Carbon::Internal
 
+// Evaluates the condition of a CHECK as a boolean value.
+//
+// This performs a contextual conversion to bool, diagnoses if the condition is
+// always true or always false, and returns its value.
+#define CARBON_INTERNAL_CHECK_CONDITION(cond) \
+  (Carbon::Internal::CheckCondition(true && (cond)))
```

Comment:
Why `true &&` instead of a cast? Or a step further, why does this need the explicit conversion versus letting the implicit conversion apply?

Reply:
Requiring an implicit conversion breaks a bunch of existing CHECKs that are being passed `std::optional`. Using an explicit conversion would allow CHECK conditions that aren't allowed in (say) an `if`. The right kind of conversion here seems to be a [contextual conversion to `bool`](https://eel.is/c++draft/conv#general-4.sentence-2) -- that's the [same conversion that is applied to the condition in an `if` or `while`](https://eel.is/c++draft/stmt.stmt#stmt.pre-6.sentence-1) -- and the easiest way to get one is by using the expression as an operand of `&&` or `||`.